Min # Vending step6

undervineg opened the pull request.

undervine committed:
Add link

undervine committed:
Adjust existing view components' positions

undervine committed:
Modify ChocoMilk to CoffeeMilk

undervine committed:
Edit whitespaces

undervine committed:
Add enum 'UserInfoKeys' for key names of NotificationCenter used when inventory updated

undervine committed:
Add NotificationCenter to variable inventory, passing poped beverage as userInfo also

undervine committed:
Add mappingImage() returning UIImage from beverage's menu type

undervine committed:
Move Stack's position

undervine committed:
Add functions attaching ImageViews of purchased beverage

undervine committed:
Add contents about step6

undervine committed:
Add struct to make ImageViews with updated positions

godrm commented on VendingMachineApp/VendingMachineApp/ProductImageMaker.swift:
```diff
@@ -0,0 +1,36 @@
+    private let size: CGSize
+    private var count: Int
+    init(_ startX: Int, _ startY: Int) {
+        self.leftTop = CGPoint(x: startX, y: startY)
```
그냥 CGPoint를 넘기면 어땠을까요?

godrm commented on VendingMachineApp/VendingMachineApp/ProductImageMaker.swift:
```diff
@@ -0,0 +1,36 @@
+        updatePosition()
+        imageView.frame = CGRect(origin: leftTop, size: size)
+        count += 1
+        return imageView
```
imageViewWithPosition() 라면 position이 매개변수로 넘어가야 할 것만 같네요.
imageView를 만드는 게 핵심이니 position을 강조할 필요는 없을 것 같아요.

godrm commented on VendingMachineApp/VendingMachineApp/ProductImageMaker.swift:
```diff
@@ -0,0 +1,36 @@
+    private func updatePosition() {
+        if count > 0 {
+            self.leftTop.x += 50
+        }
```
이 정도는 뭐 굳이 함수까지 ㅎㅎ 삼항 연산자도 충분할 것 같아요.

undervineg replied:
삼항 연산자 쓸 때, false일 때는 아무것도 안 하고 싶은데 어떻게 쓸 수 있나요?

godrm replied:
케이스마다 다른데 break 혹은 nil 만 써도 되기도 합니다.
여기서는 leftTop에 초기값을 넣으면 되지 않을까요?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -30,17 +42,43 @@ class ViewController: UIViewController {
             name: .didUpdateBalance,
             object: nil)
+        NotificationCenter.default.addObserver(
+            forName: .didUpdateInventory, object: nil, queue: nil, using: catchNotification)
```
위에서 사용한 addObserver와 여기 addObserver는 살짝 다르네요. 어떤 차이가 있는지 공부하고 댓글로 남겨보세요.

undervineg replied:
addObserver(forName:object:queue:using:)은 노티를 받고 나서 특정 작업을 수행할 수 있는데요, Notification 객체를 전달받을 수 있기 때문에 post 측에서 userInfo를 전달한 경우, userInfo 내부 데이터를 꺼내서 사용할 수 있습니다.

godrm replied:
그러면 addObserver(_:selector:name:object)는 노티를 받아서 동일하게 처리할 수 없다는 의미인가요?

undervineg replied:
아 제가 잘못 알았네요. addObserver(_:selector:name:object)도 Notification 객체를 받는 메소드를 selector에 지정해주면 Notification 객체의 UserInfo를 가지고 처리할 수 있습니다. addObserver(forName:object:queue:using:)는 셀렉터 대신 block 안에서 노티를 받은 후의 작업을 수행할 수 있는데, queue 인자를 통해 작업순서를 지정할 수 있고, block 인자는 이스케이프 클로저 형태로 되어있어 노티가 끝난 후 작업 수행이 보장돼있기 때문에 멀티쓰레드 환경에서 사용하기 좋은 형태로 이해했습니다.

godrm replied:
네 큐(queue)와 클로저를 활용해서 노티피케이션 처리를 병렬화할 수 있는 기능입니다.
항상 사용하는 함수나 API 차이점을 알고 활용하는 습관을 만들어보세요.

godrm reviewed (commented):
다음 피드백을 확인해보세요.

undervine committed:
Modify initialzer to get CGPoint and CGSize as parameter

Remove updatePosition() function.
Modify class name 'ProductImageMaker' to 'ProductImageViewMaker'.
Modify function name 'imageViewWithPosition' to 'produce'.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -16,14 +16,16 @@ class ViewController: UIViewController {
-        self.purchasedImageViewPositionY = 686
-        self.imageMaker = ProductImageMaker(purchasedImageViewPositionX, purchasedImageViewPositionY)
+        self.purchasedImageViewPosition = CGPoint(x: 40, y: 686)
+        self.purchasedImageViewSize = CGSize(width: 140, height: 100)
```
point와 size 두 가지 정보는 또 CGRect 라는 구조체로 표현할 수 있습니다. 
View 정보는 CGRect로 보관해야 하는 경우가 많아질 거에요. 
우리가 레벨2에서 좌표계산기를 배웠던 배경도 관련이 있습니다. 참고하세요

undervineg replied:
아 그렇네요. 수정하겠습니다.

undervine committed:
Modify addObserver function

undervine committed:
Replace CGPoint and CGSize to CGRect

godrm commented:
수고하셨습니다~ 다음 단계로 갑시다
